Adds UUIDs to snapshots

abeyad opened the pull request:
This commit adds a UUID for each snapshot, in addition to the already
existing repository and snapshot name. The addition of UUIDs will enable
more robust handling of the deletion of previous snapshots and lingering
files from partially failed delete operations, on top of being able to
uniquely track each snapshot.

Relates #18156

abeyad commented:
@s1monw @ywelsch your feedback would be appreciated.

ywelsch commented on core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1224,13 +1282,11 @@ public RepositoriesService getRepositoriesService() {
-            this.cause = cause;
-            this.name = name;
-            this.repository = repository;
+        public SnapshotRequest(String cause, Snapshot snapshot) {
```
While we are changing this constructor, can you put Snapshot as first argument?

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/TransportNodesSnapshotsStatus.java:
```diff
@@ -192,13 +192,13 @@ public NodeRequest() {
         public void readFrom(StreamInput in) throws IOException {
             super.readFrom(in);
-            snapshotIds = in.readList(SnapshotId::readSnapshotId);
+            snapshotIds = in.readList(streamInput -> new SnapshotId(streamInput));
```
Simpler: `snapshotIds = in.readList(SnapshotId::new);`
No need to create lambda :-)

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/TransportSnapshotsStatusAction.java:
```diff
@@ -197,20 +198,20 @@ private SnapshotsStatusResponse buildResponse(SnapshotsStatusRequest request, Li
-                    SnapshotId snapshotId = new SnapshotId(request.repository(), snapshotName);
-                    if (currentSnapshotIds.contains(snapshotId)) {
+                    Snapshot snapshot = new Snapshot(request.repository(), snapshotName);
+                    if (currentSnapshotIds.contains(snapshot)) {
```
This always returns false as `currentSnapshotIds` is of type `Set<SnapshotId>` and `snapshot` is of type `Snapshot`. Is this untested?

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/TransportSnapshotsStatusAction.java:
```diff
@@ -197,20 +198,20 @@ private SnapshotsStatusResponse buildResponse(SnapshotsStatusRequest request, Li
                         continue;
                     }
-                    SnapshotInfo snapshot = snapshotsService.snapshot(snapshotId);
+                    SnapshotInfo snapshotInfo = snapshotsService.snapshot(snapshot);
```
As we don't know the snapshot uuid here, we re-resolve the list of snapshots for every entry in `request.snapshots()`. Can we do that only once?

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/metadata/SnapshotId.java:
```diff
@@ -19,37 +19,78 @@
+     * @return snapshot id
+     */
+    public static SnapshotId get(final Snapshot snapshot, final String uuid) {
+        return new SnapshotId(snapshot, uuid);
```
As this is just a public wrapper for new Snapshot(...), I prefer that we make the constructor public and get rid of this method.

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/metadata/SnapshotId.java:
```diff
@@ -19,37 +19,78 @@
+     * @return snapshot id
+     */
+    public static SnapshotId createNew(final Snapshot snapshot) {
+        return new SnapshotId(snapshot, UUIDs.randomBase64UUID());
```
same comment here as for `get`

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/metadata/SnapshotId.java:
```diff
@@ -19,37 +19,78 @@
+public class SnapshotId implements Writeable {
 
-    private String repository;
+    public static final String UNASSIGNED_UUID = ""; // empty string so its compatible with old blob naming
```
I wonder if we should use something like `"_na_"` similar to the `INDEX_UUID_NA_VALUE`. That is more an explicit value.

ywelsch replied:
also, reference `blobId()` method in this comment

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/metadata/SnapshotId.java:
```diff
@@ -19,37 +19,78 @@
+ * SnapshotId - snapshot (repository name + snapshot name) + snapshot UUID
  */
-public class SnapshotId implements Streamable {
+public class SnapshotId implements Writeable {
```
make class final

abeyad replied:
Unfortunately I can't.  I tried it earlier, but mockito uses this class in some tests and mockito can't work with `final` classes

ywelsch replied:
I wonder why we're mocking this simple class. Maybe we can fix the tests? :-)

ywelsch commented on core/src/main/java/org/elasticsearch/common/Strings.java:
```diff
@@ -37,6 +37,7 @@
 import java.util.Iterator;
 import java.util.LinkedList;
 import java.util.List;
+import java.util.Objects;
```
remove import?

abeyad replied:
This import is removed in another PR I did on cleaning up the `Strings` class

ywelsch commented on core/src/main/java/org/elasticsearch/repositories/Repository.java:
```diff
@@ -139,4 +140,14 @@
 
+    /**
+     * Resolves snapshot name(s) in the repository to {@link SnapshotId}(s).  If any snapshots
+     * are not found, a {@link SnapshotMissingException} is thrown.
```
This reads a bit funny: "If any .. are not found"...

ywelsch commented on core/src/main/java/org/elasticsearch/repositories/Repository.java:
```diff
@@ -139,4 +140,14 @@
+     * @return snapshot ids
+     * @throws SnapshotMissingException if a snapshot is not found
+     */
+    List<SnapshotId> resolveSnapshots(String... snapshotNames);
```
maybe call this "resolveSnapshotNames"?
I would also prefer to use `List<String> snapshotNames` as parameter to bring it closer to the return type.

abeyad replied:
I struggled with this one myself, the reason I used `String...` is because it makes the API easier for when we only want to resolve a single snapshot name.  But it is quite trivial to create a list out of a single name, so I will go with what you suggested for the sake of consistency.

ywelsch commented on core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -268,20 +270,21 @@ public void initializeSnapshot(SnapshotId snapshotId, List<String> indices, Meta
+            final String snapshotName = snapshotId.getName();
+            if (snapshotFormat.exists(snapshotsBlobContainer, snapshotId.blobId()) ||
+                    snapshotLegacyFormat.exists(snapshotsBlobContainer, snapshotName)) {
+                throw new InvalidSnapshotNameException(snapshotId.getSnapshot(), "snapshot with such name already exists");
```
is it now possible to write two snapshots with same name? Is there a test for this?

ywelsch commented on core/src/main/java/org/elasticsearch/snapshots/SnapshotInfo.java:
```diff
@@ -86,23 +89,24 @@
 
-    public SnapshotInfo(String name, List<String> indices, long startTime) {
-        this(name, indices, SnapshotState.IN_PROGRESS, null, Version.CURRENT, startTime, 0L, 0, 0, Collections.emptyList());
+    public SnapshotInfo(String name, String uuid, List<String> indices, long startTime) {
```
Let's use `SnapshotId` instead here.

abeyad replied:
`SnapshotInfo` is only used within a repository, for example in a GET request: `/_snapshot/my_repo` will return snapshot infos.  Also, `SnapshotInfo` is the structure that is written to the snapshot metadata blob, which again lives in a repository already.  Adding the repository info will also give us another backward compatibility issue to worry about due to the structure being serialized to a blob in the repository.

ywelsch commented on core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -669,4 +724,45 @@ public static String testBlobPrefix(String seed) {
+
+    // TODO: this will go away once readSnapshotsList uses the index file instead of listing blobs
+    // to know which snapshots are part of a repository.  See #18156
+    private Tuple<String, String> parseNameUUIDFromBlobName(final String str) {
```
make this method static and package-visible and test it thoroughly?

abeyad replied:
Its a good idea, though this method is disappearing in the very next PR (I already have it implemented), where we do not depend on blob listing and parsing the UUID from the blob name.  So perhaps thoroughly testing this method in particular won't be a good use of time as it will disappear very soon thereafter.  Also, the other tests have captured issues with parsing the blob names that I've uncovered and fixed in this method.  So its not entirely untested.

ywelsch commented on core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -586,6 +612,16 @@ protected void writeSnapshotList(List<SnapshotId> snapshots) throws IOException
+            if (snapshotsBlobContainer.blobExists(SNAPSHOTS_FILE) == false) {
+                return Collections.emptyList();
+            }
+        } catch(UnsupportedOperationException e){
```
space missing after catch

ywelsch commented on core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -586,6 +612,16 @@ protected void writeSnapshotList(List<SnapshotId> snapshots) throws IOException
      */
     protected List<SnapshotId> readSnapshotList() throws IOException {
+        try {
+            if (snapshotsBlobContainer.blobExists(SNAPSHOTS_FILE) == false) {
```
why do we need to add this here? Can you explain please? Also I'm wondering why blobExists is not supported by URLRepository (Shouldn't that just correspond to a HEAD request to the resource?).

abeyad replied:
First, we need to check if the repository has a snapshot index file.  If it is a fresh repository, it won't have that snapshot index file and the `readBlob` call just below will throw exceptions.

abeyad replied:
Regarding URL Repository having blobExists, I suppose you are right, it could just be a HEAD request.  I'm not sure why its not done, but its worth implementing or trying to and see what issues arise (in a separate PR).

abeyad replied:
Just to follow up, you are correct, we do not need this check here, because as it stands, `readSnapshotList` is never invoked before it is written.  In subsequent work, we will probably have to re-introduce this check, but I'm removing it from this PR for now as it is not necessary for it.

ywelsch commented on core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -594,10 +630,29 @@ protected void writeSnapshotList(List<SnapshotId> snapshots) throws IOException
+                                        }
+                                        snapshots.add(SnapshotId.get(new Snapshot(repositoryName, name), uuid));
+                                    }
+                                    // the old format pre 5.0 that only contains the snapshot name, use the name as the uuid too
```
do we have BWC tests for this?

ywelsch commented on core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -669,4 +724,45 @@ public static String testBlobPrefix(String seed) {
+            if (resolvedSnapshotId == null) {
+                throw new SnapshotMissingException(new Snapshot(repositoryName, snapshotName));
+            }
+            ids.add(resolvedSnapshotId);
```
check that size of list that is returned equals size of parameters?

ywelsch commented on core/src/main/java/org/elasticsearch/snapshots/RestoreService.java:
```diff
@@ -940,7 +937,7 @@ public String cause() {
          */
         public String name() {
-            return name;
+            return snapshot.getName();
```
we should probably expose snapshot directly as I see that usage points create a new Snapshot(..) again based on the data here...

ywelsch commented on core/src/main/java/org/elasticsearch/snapshots/SnapshotException.java:
```diff
@@ -30,21 +29,21 @@
         super(in);
         if (in.readBoolean()) {
-            snapshot = SnapshotId.readSnapshotId(in);
+            snapshot = new Snapshot(in);
```
Remove if/else by `in.readOptionalWriteable(Snapshot::new)`

ywelsch commented on core/src/main/java/org/elasticsearch/snapshots/SnapshotInfo.java:
```diff
@@ -443,12 +463,18 @@ public static SnapshotInfo fromXContent(final XContentParser parser) throws IOEx
-        return new SnapshotInfo(name, indices, state, reason, version, startTime, endTime, totalShard, successfulShards, shardFailures);
+        if (uuid == null) {
+            // the old format where there wasn't a UUID
+            uuid = "";
```
use the constant defined in SnapshotId...

ywelsch commented on core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -474,30 +477,31 @@ private SnapshotInfo inProgressSnapshot(SnapshotsInProgress.Entry entry) {
      */
-    public Map<ShardId, IndexShardSnapshotStatus> snapshotShards(SnapshotId snapshotId) throws IOException {
-        validate(snapshotId);
+    public Map<ShardId, IndexShardSnapshotStatus> snapshotShards(Snapshot snapshot) throws IOException {
```
Pass SnapshotInfo directly to this method? See `TransportSnapshotsStatusAction`. No need to revalidate...

ywelsch commented on core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -582,7 +586,7 @@ public ClusterState execute(ClusterState currentState) throws Exception {
                         } else if (snapshot.state() == State.INIT && newMaster) {
                             // Clean up the snapshot that failed to start from the old master
-                            deleteSnapshot(snapshot.snapshotId(), new DeleteSnapshotListener() {
+                            deleteSnapshot(snapshot.snapshotId().getSnapshot(), new DeleteSnapshotListener() {
```
why reresolve again here?

ywelsch commented on core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -868,11 +872,24 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+     * @param snapshot   snapshot
      * @param listener   listener
      */
+    public void deleteSnapshot(final Snapshot snapshot, final DeleteSnapshotListener listener) {
```
this method is dangerous as it wrong usage leads to re-resolving all the time. Maybe just remove it and do the `resolveSnapshotId` in TransportDeleteSnapshotAction

ywelsch commented on core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -868,11 +872,24 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
     public void deleteSnapshot(final SnapshotId snapshotId, final DeleteSnapshotListener listener) {
-        validate(snapshotId);
+        final Snapshot snapshot = snapshotId.getSnapshot();
+        validate(snapshot);
```
no need to validate? can't we assume that once we have a SnapshotId that validation has happened?

ywelsch commented on core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1159,6 +1177,48 @@ public RepositoriesService getRepositoriesService() {
+     * @return the snapshot ids
+     * @throws SnapshotMissingException if a snapshot could not be found
+     */
+    public List<SnapshotId> resolveSnapshotIds(final String repositoryName, final Set<String> snapshotNames) {
```
This method is only used by the single collection version above. Do you see other usages?

ywelsch commented on core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1159,6 +1177,48 @@ public RepositoriesService getRepositoriesService() {
+        if (snapshots.isEmpty() == false) {
+            ids.addAll(repository.resolveSnapshots(snapshots.toArray(new String[snapshots.size()])));
+        }
+        return Collections.unmodifiableList(ids);
```
Are we guaranteed that the returned list has same size as parameter list?

ywelsch commented on plugins/repository-azure/src/test/java/org/elasticsearch/cloud/azure/storage/AzureStorageServiceMock.java:
```diff
@@ -94,8 +94,15 @@ public OutputStream getOutputStream(String account, LocationMode mode, String co
+            final String checkBlob;
+            if (keyPath != null) {
+                // strip off key path from the beginning of the blob name
+                checkBlob = blobName.replace(keyPath, "");
```
can you explain why this change was needed?

abeyad replied:
Because the blobNames are named with the directory path preceding the blob name.  So a blob named `foobar` in the repository will actually have a blob name of `/some/azure/path/foo`.  This causes issues because if I wanted to list all blob names in the mock repository that start with `foo`, none would match because they all start with the key path (i.e. `/some/azure/path`).  This basically makes it easier to work with the blob names in this mock.  Its worth revisiting whether we need the key path at all to be stored in the blob names for this mock repository, but I think that's a separate issue that needs a separate PR.

abeyad replied:
@ywelsch to further elaborate, the lack of taking the `keyPath` into account when listing blobs was a bug in`AzureStorageServiceMock`.  The reason this never manifested before is the `listBlobsByPrefix` was only called from `BlobStoreRepository#snapshots()`, which in turn was only called during `initializeSnapshot` and `finalizeSnapshot` to get the current list of snapshots in the index file to add to it.  This means the snapshot file did not have the correct entries, but it didn't matter because it was never used anyway (because `listBlobsByPrefix` was always invoked instead).  However, because now we are also using the `BlobStoreRepository#snapshots()` method in `resolveSnapshotNames`, it has manifested this issue in the `AzureStorageServiceMock`.  Hence the change above.

ywelsch replied:
ok, sounds good.

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java:
```diff
@@ -74,32 +73,29 @@ protected ClusterBlockException checkBlock(GetSnapshotsRequest request, ClusterS
-                snapshotInfoBuilder.addAll(snapshotsService.currentSnapshots(request.repository()));
+                snapshotInfoBuilder.addAll(snapshotsService.currentSnapshots(repository));
             } else {
-                Set<String> snapshotsToGet = new LinkedHashSet<>(); // to keep insertion order
```
why did you remove this?
What if we have multiple pattern that match multiple snapshots? we now add these multiple times.
For example:
`foo*,*bar` will match the snapshot `foobar` twice.

abeyad replied:
Good point, I didn't think of that.  I'll re-add it.  Also, seems like no tests exist for that case so I'll add it.

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/RestoreInProgress.java:
```diff
@@ -77,12 +78,14 @@ public RestoreInProgress(Entry... entries) {
      * @return restore metadata or null
      */
-    public Entry snapshot(SnapshotId snapshotId) {
+    public Entry snapshot(final String repositoryName, final String snapshotName) {
```
IDEA says that this method is not used... maybe just remove it.

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java:
```diff
@@ -348,9 +348,10 @@ public SnapshotsInProgress(Entry... entries) {
     }
 
-    public Entry snapshot(Snapshot snapshot) {
+    public Entry snapshot(final String repositoryName, final String snapshotName) {
```
we can make this `public Entry snapshot(Snapshot snapshot)`
The only place where there is an issue is in the tests. There, we can manually iterate over the entries...

ywelsch replied:
I also wonder if we should make `entries` a set (follow up).

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/metadata/Snapshot.java:
```diff
@@ -17,37 +17,41 @@
  */
 
-package org.elasticsearch.snapshots;
+package org.elasticsearch.cluster.metadata;
```
I prefer to leave Snapshot in snapshots package.

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/metadata/Snapshot.java:
```diff
@@ -78,18 +82,29 @@ public boolean equals(Object o) {
+    }
+
+    /**
+     * Returns whether the snapshot logically equivalent to the given repository name + snapshot name combination.
```
word missing

ywelsch replied:
I would just say: "whether it has the same name and repository".

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/metadata/Snapshot.java:
```diff
@@ -78,18 +82,29 @@ public boolean equals(Object o) {
+    /**
+     * Returns whether the snapshot logically equivalent to the given repository name + snapshot name combination.
+     */
+    public boolean isSame(final String repositoryName, final String snapshotName) {
```
this is effectively used only in 3 places (due to weird listener completion logic). maybe we can just do equals there and get rid of this method.

ywelsch commented on core/src/main/java/org/elasticsearch/rest/action/cat/RestSnapshotAction.java:
```diff
@@ -79,7 +79,7 @@ protected void documentation(StringBuilder sb) {
         return new Table()
                 .startHeaders()
-                .addCell("id", "alias:id,snapshotId;desc:unique snapshot id")
+                .addCell("id", "alias:snapshot,snapshot;desc:unique snapshot")
```
you probably intended to write "alias:id,snapshot".

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/create/TransportCreateSnapshotAction.java:
```diff
@@ -87,16 +85,16 @@ public void onResponse() {
-                        public void onSnapshotCompletion(SnapshotId snapshotId, SnapshotInfo snapshotInfo) {
-                            if (snapshot.equals(snapshotId.getSnapshot())) {
+                        public void onSnapshotCompletion(Snapshot snapshot, SnapshotInfo snapshotInfo) {
+                            if (snapshot.isSame(request.repository(), request.snapshot())) {
```
The listeners in SnapshotService and RestoreService could just themselves implement this matching logic. The API would then be 

```
snapshotsService.addListener(repository, snapshotName, new SnapshotsService.SnapshotCompletionListener() { ... });
```

abeyad replied:
Its not clear to me what the above API is intended to achieve?  As I see it, the listeners have this check in them because on snapshot/restore completion, every registered listener is looped through and `onCompletion` is called, so the check is to make sure the listener only truly runs on the right snapshot.  I think the listeners could be redone to be in a map instead of a list, where the map has entries e.g. `Snapshot -> List<SnapshotCompletionListener>`.  But that change seems beyond the scope of this PR.

ywelsch commented on core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -193,9 +196,11 @@ public SnapshotInfo snapshot(Snapshot snapshot) {
+        final String repositoryName = request.repositoryName;
+        final String snapshotName = request.snapshotName;
+        validate(repositoryName, snapshotName);
+        final SnapshotId snapshotId = new SnapshotId(snapshotName); // creates a new UUID for the snapshot
```
I prefer that we make this explicit by passing `UUIDs.randomBase64UUID()` here and removing the overloaded constructor.

ywelsch commented on core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -204,36 +209,44 @@ public void createSnapshot(final SnapshotRequest request, final CreateSnapshotLi
                 validate(request, currentState);
 
+                // check if the snapshot name already exists in the repository
+                if (snapshotNameExists(repositoryName, snapshotName)) {
```
This does IO in the cluster state update thread 👎 Super dangerous.
We should probably put assertions in BlobStoreRepository that we are not the cluster state update thread to help prevent mistakes like this.

abeyad commented:
@ywelsch Commit https://github.com/elastic/elasticsearch/pull/18228/commits/92df39f33a3055a4af69731ae0d67d47501d32a2 addresses your code review comments and adds some tests.  Commit https://github.com/elastic/elasticsearch/pull/18228/commits/077fafa7731282d9b5e2e64ee834cbea1d2c3493 adds more tests.

ywelsch commented on core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -199,7 +199,12 @@ public void createSnapshot(final SnapshotRequest request, final CreateSnapshotLi
+        final SnapshotId snapshotId = new SnapshotId(snapshotName, UUIDs.randomBase64UUID()); // new UUID for the snapshot
+
+        // check if the snapshot name already exists in the repository
+        if (snapshotNameExists(repositoryName, snapshotName)) {
```
I think this check needs to go to `beginSnapshot` (where snapshot has officially become part of cluster state) to ensure that no two snapshots with same name are created in parallel. Where was this originally checked?

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java:
```diff
@@ -74,32 +75,41 @@ protected ClusterBlockException checkBlock(GetSnapshotsRequest request, ClusterS
+                        if (snapshotInfos.containsKey(snapshotOrPattern)) {
+                            continue; // already got the snapshot info for this snapshot
+                        }
+                        List<SnapshotId> snapshotIds = snapshotsService.resolveSnapshotNames(repository,
```
This leads to resolving the snapshot name on the repository for each entry in the list, which is very wasteful (We query the list of snapshots n times and thus have n separate network requests on S3/GCE etc.). This method should be rewritten so that we query the snapshot names on the repository once and then get all the snapshotInfo elements for the entries that we requested in a subsequent step.

abeyad replied:
Good point, i thought i had fixed that, i'll amend this and push up a fix.

ywelsch commented on core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1159,9 +1219,41 @@ public RepositoriesService getRepositoriesService() {
+        Objects.requireNonNull(repositoryName);
+        Objects.requireNonNull(snapshotNames);
+        List<SnapshotId> ids = new ArrayList<>();
+        Set<String> snapshots = new LinkedHashSet<>(snapshotNames);
```
what if parameter list contains duplicates?

ywelsch commented on core/src/test/java/org/elasticsearch/repositories/blobstore/BlobStoreRepositoryTests.java:
```diff
@@ -0,0 +1,184 @@
+                                       .get();
+        final SnapshotId snapshotId2 = createSnapshotResponse.getSnapshotInfo().snapshotId();
+
+        logger.info("--> make sure each node's repository can resolve the snapshots");
```
there is just 1 node

ywelsch commented on core/src/test/java/org/elasticsearch/repositories/blobstore/BlobStoreRepositoryTests.java:
```diff
@@ -0,0 +1,184 @@
+        assertThat(putRepositoryResponse.isAcknowledged(), equalTo(true));
+
+        final RepositoriesService repositoriesService = getInstanceFromNode(RepositoriesService.class);
+        @SuppressWarnings("unchecked") final FsRepository repository = (FsRepository) repositoriesService.repository(repositoryName);
```
shouldn't we cast just to blobstorerepository?

ywelsch commented on core/src/test/java/org/elasticsearch/repositories/blobstore/BlobStoreRepositoryTests.java:
```diff
@@ -0,0 +1,184 @@
+        assertThat(readSnapshotIds, equalTo(snapshotIds));
+    }
+
+    public void testParseUUIDFromBlobName() {
```
lets put this into a unit tests class

ywelsch commented on core/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java:
```diff
@@ -2203,13 +2195,173 @@ public void testListCorruptedSnapshot() throws Exception {
+                                                       .put("chunk_size", randomIntBetween(100, 1000), ByteSizeUnit.BYTES)));
+
+        logger.info("--> get snapshots on an empty repository");
+        /*expectThrows(SnapshotMissingException.class, () -> client.admin()
```
remove?

abeyad replied:
Thanks for alerting me, that actually belongs, I uncommented it out.

ywelsch commented:
Left a few more comments. We're getting close...

abeyad commented:
@ywelsch I experimented with removing `resolveSnapshotNames` in https://github.com/elastic/elasticsearch/pull/18228/commits/69bbdcc1b413ef39bac782aef2d6364d4ceaa90d

WDYT?

abeyad commented:
@ywelsch you may also be interested in this not-so-pleasant commit https://github.com/elastic/elasticsearch/pull/18228/commits/d196050aff3f5000240dfe50de5df1ed2709021b :)

imotov commented on core/src/test/java/org/elasticsearch/snapshots/AbstractSnapshotIntegTestCase.java:
```diff
@@ -107,18 +106,29 @@ public void waitForBlock(String node, String repository, TimeValue timeout) thro
+                    boolean found = false;
+                    for (SnapshotsInProgress.Entry entry : snapshotsInProgress.entries()) {
+                        final Snapshot curr = entry.snapshot();
+                        if (curr.getRepository().equals(repository) && curr.getSnapshotId().getName().equals(snapshotName)) {
```
I have seen this comparison in several places. Maybe, it makes sense to move inside Snapshot and have something like `boolean hasName(String repository, String name)` method where it will take place? Not sure if `hasName` is a good name for this method though.

imotov commented on core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -71,6 +71,8 @@
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
+import java.util.function.Predicate;
```
Leftovers?

abeyad replied:
Yes, correct, I just removed it.

imotov commented on core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -192,8 +212,11 @@ public SnapshotInfo snapshot(SnapshotId snapshotId) {
+        final String repositoryName = request.repositoryName;
+        final String snapshotName = request.snapshotName;
+        validate(repositoryName, snapshotName);
+        final SnapshotId snapshotId = new SnapshotId(snapshotName, UUIDs.randomBase64UUID()); // new UUID for the snapshot
```
We call `UUIDs.randomBase64UUID()` in 17 different places in this PR. Maybe move into a SnapshotId constructor or a static create(String name) method?

ywelsch replied:
It's used 1 time in production code and 16 times in the tests... For most of these tests, we also chose a random snapshot name and when chose a specific snapshot name we can as well just use a fixed uuid. I'm not sure that warrants adding a constructor...

imotov commented on core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -292,10 +330,20 @@ private static void validate(SnapshotId snapshotId) {
         try {
-            Repository repository = repositoriesService.repository(snapshot.snapshotId().getRepository());
+            // check if the snapshot name already exists in the repository
+            if (snapshotNameExists(snapshot.snapshot())) {
```
I think this check should go into initializeSnapshot or repository.

abeyad replied:
Do you feel that way because the check only looks at what is in the repository, so it belongs as a repository level method?

imotov replied:
There are a few issues here. First I think we have a race condition here. Moving this check to the the repository level will not remove it right now, but will tighten the loop. Hopefully, we will solve it there at some point of time with tightening the blob store api and adding locks. The second issue is that it can be done more effectively on the repository level.

abeyad replied:
I've been trying to capture some of the race conditions I've encountered here: https://github.com/elastic/elasticsearch/issues/18551

But I think there are a few race conditions in general, and one of them relates to the fact that we have two separate places where we are retrieving snapshot info without a lock (cluster state and repository), so at any given time, the state could change and for example, a snapshot can move from in progress (cluster state) to completed (in repository), and we could double count if we don't properly synchronize around these snapshot related operations.

ywelsch commented on core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -244,41 +270,53 @@ public TimeValue timeout() {
 
     /**
+     * Checks if a snapshot by the same name exists in the repository, either
+     * currently in progress or persisted in the repository.
```
It only checks those persisted in the repository...

abeyad replied:
Yes, good catch, outdated java doc comments from how i previously implemented it.

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java:
```diff
@@ -72,34 +78,55 @@ protected ClusterBlockException checkBlock(GetSnapshotsRequest request, ClusterS
+                    } else {
+                        return s1;
+                    }
+                }));
```
Maybe simpler just to do it the oldschool way.

```
final Map<String, SnapshotId> allSnapshotIds = new HashMap<>();
for (SnapshotInfo snapshotInfo : snapshotsService.currentSnapshots(repository)) {
    SnapshotId snapshotId = snapshotInfo.snapshotId();
    allSnapshotIds.put(snapshotId.getName(), snapshotId);
}
for (SnapshotId snapshotId : snapshotsService.snapshotIds(repository)) {
    allSnapshotIds.put(snapshotId.getName(), snapshotId);
}
```

abeyad replied:
I agree, it became cumbersome b/c of having to handle the fact that `Collectors#toMap` requires a merge strategy for duplicate keys (hence the third lamda parameter).

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/TransportSnapshotsStatusAction.java:
```diff
@@ -190,41 +196,49 @@ private SnapshotsStatusResponse buildResponse(SnapshotsStatusRequest request, Li
+            final Set<String> requestedSnapshotNames = Sets.newHashSet(request.snapshots());
+            final Set<String> currentSnapshotNames = currentSnapshotIds.stream().map(SnapshotId::getName).collect(Collectors.toSet());
+            final Map<String, SnapshotId> matchedSnapshotIds = snapshotsService.snapshotIds(repositoryName).stream()
+                .filter(s -> requestedSnapshotNames.contains(s.getName()) && currentSnapshotIds.contains(s) == false)
```
I wonder if we can simplify this. What we currently achieve here is that we prioritize running snapshots over stored ones (with same name). If that's what we want, we can instead of creating `currentSnapshotIds` at the beginning of the method create a `Set<String> currentSnapshotNames`. Instead of `matchedSnapshotIds` we just use `Map<String, SnapshotId> storedSnapshotIds` which are just all  the stored snapshots. Using the check `if (currentSnapshotNames.contains(snapshotName)) {` below we then get the same behavior as now?

ywelsch commented on core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -864,15 +925,37 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
 
     /**
+     * Deletes a snapshot from the repository, looking up the {@link Snapshot} reference before deleting.
+     * If the snapshot is still running cancels the snapshot first and then deletes it from the repository.
```
that's not whats happening here. If a snapshot with the name exists in the repository only the snapshot in the repository is deleted. The running snapshot is not canceled. We should do both imo.

abeyad replied:
I can do both, because while it is the behavior that *if* there are two snapshots of the same name, one of them must be in the cluster state as "in progress" that is in the process of being removed from the cluster state, we can add the removal here as a safeguard because in `remoteSnapshotFromClusterState`, if the snapshot is not found in the cluster state (i.e. the `deleteSnapshot` removed it), then its essentially a no-op.

abeyad replied:
@ywelsch and I discussed this; because two different threads will submit cluster state update tasks to remove the current snapshot from the cluster state, and one of them would try to delete physical contents on disk (which don't exist) via `clusterStateProcessed` callback, and since we are guaranteed to have the "in-progress" snapshot removed by `removeSnapshotFromClusterState` (because the snapshot name exists check will fail), we will leave it as is.

ywelsch commented on core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -120,48 +123,65 @@ public SnapshotsService(Settings settings, ClusterService clusterService, IndexN
+            currentSnapshots(repositoryName, snapshotIds.stream().map(SnapshotId::getName).collect(Collectors.toList()));
         for (SnapshotsInProgress.Entry entry : entries) {
             snapshotSet.add(inProgressSnapshot(entry));
+            snapshotIdsToIterate.remove(entry.snapshot().getSnapshotId());
```
here, we prioritize matching running snapshots to stored snapshots, but do it the other way around when we explicitly match in TransportGetSnapshotsAction?

abeyad replied:
So, this is not exactly true.  In `TransportGetSnapshotsAction`, `TransportSnapshotsStatusAction`, and in `SnapshotsService#snapshots`, we first look at the current in progress snapshots, then at the repository.  This is necessary to avoid looking at the repository for snapshots that don't exist in it (thereby throwing exceptions).  Also, in each of these, the idea is that *if* there are two snapshots of the same name, the one from the repository should override it, given that we know the one "in-progress" is soon to be deleted from the cluster state due to the naming conflict.  However, `TransportSnapshotsStatusAction` did not do this, due to the line of checking if the snapshot is in currentSnapshotIds and if so, skips over it.  Instead I changed it as can be found in the latest commit here: https://github.com/elastic/elasticsearch/pull/18228/commits/a17039c86d7d9bfd10f77bbb15ca1b7a43496226#diff-9426f24b722fd1c3664ad17ead551f83L211

This leaves `deleteSnapshot` to have the only "inconsistent" behavior... but I believe it doesn't make sense to look first in current, only to know we have to look then into the repository and if found in the repository, discard the one in current, for the purposes of deletion.

abeyad commented:
@ywelsch @imotov The latest commit addressing code review comments is here: https://github.com/elastic/elasticsearch/pull/18228/commits/a17039c86d7d9bfd10f77bbb15ca1b7a43496226

Also, @imotov has requested I create an infrastructure for testing repositories with old snapshots interoperating with new snapshots.  For now, the only thing to test is the blob names with respect to the changes, but it will make it easier to test changes down the road with generational index files.

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java:
```diff
@@ -72,34 +76,47 @@ protected ClusterBlockException checkBlock(GetSnapshotsRequest request, ClusterS
             List<SnapshotInfo> snapshotInfoBuilder = new ArrayList<>();
             if (isAllSnapshots(request.snapshots())) {
-                snapshotInfoBuilder.addAll(snapshotsService.snapshots(request.repository(), request.ignoreUnavailable()));
+                snapshotInfoBuilder.addAll(snapshotsService.snapshots(repository, snapshotsService.snapshotIds(repository),
```
This only loads the snapshots from disk.
Please fix and add a test.

ywelsch commented on core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -120,48 +123,66 @@ public SnapshotsService(Settings settings, ClusterService clusterService, IndexN
-        List<SnapshotsInProgress.Entry> entries = currentSnapshots(repositoryName, null);
+    public List<SnapshotInfo> snapshots(final String repositoryName, List<SnapshotId> snapshotIds, final boolean ignoreUnavailable) {
+        final Set<SnapshotInfo> snapshotSet = new HashSet<>();
+        final List<SnapshotId> snapshotIdsToIterate = new ArrayList<>(snapshotIds);
```
We can use a set here instead (it's sorted at the end anyhow).

```
final Set<SnapshotId> snapshotIdsToIterate = new HashSet<>(snapshotIds);
        // first, look at the snapshots in progress
        final List<SnapshotsInProgress.Entry> entries =
            currentSnapshots(repositoryName, snapshotIds.stream().map(SnapshotId::getName).collect(Collectors.toList()));
        for (SnapshotsInProgress.Entry entry : entries) {
            snapshotSet.add(inProgressSnapshot(entry));
            snapshotIdsToIterate.remove(entry.snapshot().getSnapshotId());
        }
```

abeyad commented:
@ywelsch @imotov 

This commit addresses the `TransportGetSnapshotsAction` issue with tests: https://github.com/elastic/elasticsearch/pull/18228/commits/ff262aecd28b8a7f92a2545901a40458dbb55713
This commit adds testing for repository interoperability with previous versions: https://github.com/elastic/elasticsearch/pull/18228/commits/1bbd2a52ebf48bfea32fdd99f9dcc6d35ca4f71e

imotov commented on core/src/test/java/org/elasticsearch/bwcompat/MixedRepositoryInteroperabilityIT.java:
```diff
@@ -0,0 +1,178 @@
+/**
+ * Tests that a repository can handle snapshots of the new (5.0) and old (pre 5.0) formats,
+ * as blob names and repository blobs have changed between the snapshot versions.
+ */
```
I think this comment has very limited shelf life :) It might make sense to omit the concrete versions since we are going to use this test after 5.0. Basically, say something about testing upgrading repository from the previous version to the new version. It might also make sense to rename the test to include word Upgraded since mixed is somewhat ambiguous here. Maybe RepositoryUpgradabilityIT?

imotov commented on core/src/test/java/org/elasticsearch/bwcompat/MixedRepositoryInteroperabilityIT.java:
```diff
@@ -0,0 +1,178 @@
+            snapshotInfosFromRepo.forEach(snapshotInfo -> {
+                if (snapshotInfo.snapshotId().getName().equals(snapshotName3)) {
+                    // only the last snapshot has all the indices
+                    assertThat(Sets.newHashSet(snapshotInfo.indices()), equalTo(indices));
```
It looks like we only check that snapshot command returned us the index name as a result of the snapshot. I think it might be better to test that we still can restore the old snapshot as well as the new snapshot.

abeyad replied:
I didn't do that because we test restoration in `RestoreBackwardsCompatIT#testRestoreOldSnapshots`

imotov replied:
Yes. But we don't test that we still can restore after we (potentially) messed things by writing a new snapshot. Maybe we didn't actually write anything, but instead wiped old snapshot up. We wouldn't know.

abeyad replied:
I see what you mean, testing restore after the repository has had new
snapshots added to it.  I'll add a test for this.

On Tue, May 31, 2016 at 4:50 PM, Igor Motov <notifications@github.com>
wrote:

> In
> core/src/test/java/org/elasticsearch/bwcompat/MixedRepositoryInteroperabilityIT.java
> <https://github.com/elastic/elasticsearch/pull/18228#discussion_r65260020>
> :
>
> > +            logger.info("--> take a new snapshot with a new index");
> > +            createIndex(indexName2);
> > +            indices.add(indexName2);
> > +            for (int i = 0; i < 10; i++) {
> > +                index(indexName2, "doc", Integer.toString(i), "foo", "new-bar-" + i);
> > +            }
> > +            refresh();
> > +            snapshotInfos.add(createSnapshot(repoName, snapshotName3));
> > +
> > +            logger.info("--> get the snapshots with the newly created snapshot [{}]", snapshotName3);
> > +            snapshotInfosFromRepo = Sets.newHashSet(getSnapshots(repoName));
> > +            assertThat(snapshotInfosFromRepo, equalTo(snapshotInfos));
> > +            snapshotInfosFromRepo.forEach(snapshotInfo -> {
> > +                if (snapshotInfo.snapshotId().getName().equals(snapshotName3)) {
> > +                    // only the last snapshot has all the indices
> > +                    assertThat(Sets.newHashSet(snapshotInfo.indices()), equalTo(indices));
>
> Yes. But we don't test that we still can restore after we (potentially)
> messed things by writing a new snapshot. Maybe we didn't actually write
> anything, but instead wiped old snapshot up. We wouldn't know.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/18228/files/d973e652a47f5a481de49d99744661f98568eea5#r65260020>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABjkQd5CtgcygmBcAvsHwF8kz2PncuUwks5qHJ8ygaJpZM4IauFn>
> .
>

imotov commented on core/src/test/java/org/elasticsearch/bwcompat/MixedRepositoryInteroperabilityIT.java:
```diff
@@ -0,0 +1,178 @@
+     * could contain a snapshot name + UUID pair(newer versions) or could contain only the snapshot name (old versions).
+     */
+    public void testSnapshotsIndexFile() throws Exception {
+
```
This test looks too short :)

abeyad replied:
Ah yes, good catch :) I added tests for the snapshot index file in https://github.com/elastic/elasticsearch/pull/18228/files/d973e652a47f5a481de49d99744661f98568eea5#diff-32116318faa60fd0d82c4abd0551f352R111

I forgot to remove this method, I will remove it now

abeyad commented:
@imotov https://github.com/elastic/elasticsearch/pull/18228/commits/9969ec3cf721ebf2f00b2fafbf76eebb9be1b071 fixes the issues with the upgradability test

imotov commented:
LGTM

abeyad commented:
Thank you for the review and valuable comments, @imotov!

ywelsch commented:
LGTM 2! Thanks for all the work here @abeyad. I think we've both learned quite a lot about some of the finer details of snap/restore.

abeyad commented:
@ywelsch thank you for your patience and great feedback!  Indeed we have learned a lot more on snapshot/resotre, which means more work ahead :)
